rfd/xxxx-azure-node-join.md

```diff
@@ -0,0 +1,154 @@
+
+## Why
+
+This will make provisioning nodes on Azure simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+In place of a join token, nodes will present a JWT access token and an attested
+data document to the auth server, both of which will be fetched on the nodes
+via instance metadata. The attested data request can include a user-specified
+nonce which will also be signed. Like the custom headers in IAM join, the nonce
+allows Teleport to issue a challenge (crypto random string of bytes) that the
+node must include in the signed request. We can use gRPC bidirectional streams
+to implement the following flow at a new `RegisterWithAzureToken` gRPC endpoint:
+
+1. The Node initiates a `RegisterWithAzureToken` request.
+2. The auth server enforces that TLS is used for transport.
+3. The auth server sends a base64 encoded 24 byte crypto-random challenge.
+  - We use 24 bytes instead of 32 because the nonce parameter is limited to 32
+    characters, and 24 bytes require 32 base64 characters.
+4. The node requests a) the attested data document using the challenge as the
+  nonce and b) an access token. Azure signs and returns both.
+5. The node sends a join request to the auth server, including the signed
+  document, token, and other parameters in a `RegisterUsingTokenRequest`.
+6. The auth server checks that attested data is valid:
+  - The document's pkcs7 signature is valid.
+  - The returned nonce matches the issued challenge.
+7. The auth server checks that the access token is valid:
+  - The access token's signature is valid.
+  - The access token was not issues before the start of the
+  `RegisterWithAzureToken` request.
+  - The requested Teleport token allows the node to join.
+8. The auth server sends credentials to join the cluster.
+
+As with IAM join, the flow will occur within a single streaming gRPC request and
+only 1 attempt with a given challenge will be allowed, with a 1 minute timeout.
+
+an Azure VM will need either a system- or user-assigned
+[managed identity](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm)
+to be able to request an access token. The identity does not require any
+particular permissions.
+
+### Teleport Configuration
+
+The existing provision token type can be extended to support Azure
+authentication, using new Azure-specific fields in the token rules section.
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: example_azure_token
+spec:
+  roles: [Node,Kube,Db]
+  azure:
+    allow:
+      # Subscription from which nodes can join. Required.
+      - azure_subscription: "22222222"
+
+        # Resource groups from which nodes can join. If empty or omitted, nodes
+        # from any resource group are allowed.
+        azure_resource_groups: ["rg1", "rg2"]
```

Comment:
Where do we fetch the resource groups from ? Is this encoded within the access token ?

Reply:
Ah, is this `xms_mirid` ?